sharding/nodesCoordinator/hashValidatorShuffler.go

```diff
@@ -285,26 +297,42 @@ func shuffleNodes(arg shuffleNodesArg) (*ResUpdateNodes, error) {
 
 	err = moveMaxNumNodesToMap(newEligible, newWaiting, arg.nodesMeta, arg.nodesPerShard)
 	if err != nil {
-		log.Warn("moveNodesToMap failed", "error", err)
+		return nil, fmt.Errorf("moveNodesToMap failed, error: %w", err)
 	}
 
-	err = distributeValidators(newWaiting, arg.newNodes, arg.randomness, false)
+	err = checkAndDistributeNewNodes(newWaiting, arg.newNodes, arg.randomness, arg.flagStakingV4Step3)
 	if err != nil {
-		log.Warn("distributeValidators newNodes failed", "error", err)
+		return nil, fmt.Errorf("distributeValidators newNodes failed, error: %w", err)
+	}
+
+	shuffledNodesCfg := &shuffledNodesConfig{
+		numShuffled:        getNumPubKeys(shuffledOutMap),
+		numNewEligible:     getNumPubKeys(newEligible),
+		numNewWaiting:      getNumPubKeys(newWaiting),
+		numSelectedAuction: uint32(len(arg.auction)),
+		maxNumNodes:        arg.maxNumNodes,
+		flagStakingV4Step2: arg.flagStakingV4Step2,
 	}
 
 	if arg.flagStakingV4Step3 {
+		log.Debug("distributing selected nodes from auction to waiting",
+			"num auction nodes", len(arg.auction), "num waiting nodes", shuffledNodesCfg.numNewWaiting)
+
 		// Distribute selected validators from AUCTION -> WAITING
-		err = distributeValidators(newWaiting, arg.auction, arg.randomness, false)
+		err = distributeValidators(newWaiting, arg.auction, arg.randomness, arg.flagBalanceWaitingLists)
 		if err != nil {
-			log.Warn("distributeValidators auction list failed", "error", err)
+			return nil, fmt.Errorf("distributeValidators auction list failed, error: %w", err)
 		}
 	}
-	if !arg.flagStakingV4Step2 {
+
+	if shouldDistributeShuffledToWaiting(shuffledNodesCfg) {
+		log.Debug("distributing shuffled out nodes to waiting",
+			"num shuffled nodes", shuffledNodesCfg.numShuffled, "num waiting nodes", shuffledNodesCfg.numNewWaiting)
+
 		// Distribute validators from SHUFFLED OUT -> WAITING
 		err = arg.distributor.DistributeValidators(newWaiting, shuffledOutMap, arg.randomness, arg.flagBalanceWaitingLists)
 		if err != nil {
-			log.Warn("distributeValidators shuffledOut failed", "error", err)
+			return nil, fmt.Errorf("distributeValidators shuffled out failed, error: %w", err)
```

Comment:
also backwards incompatibility here?

Reply:
This is related to my comment from [here](https://github.com/multiversx/mx-chain-go/pull/5061#discussion_r1134060699).
An error here could never happend, that's why it was only treated as a warn;
This case never manifested on mainnet anyway, it's safe to refactor it